### DISCUSSION TITLE
Why is the return type of `useWatch` a deep partial unlike the one from `useFormContext`'s `watch`?

### DISCUSSION THREAD

**Discussion by zetdotdev** [react-hook-form/react-hook-form#10610] [Q&A]
Using `useWatch<T>()` (parameterless) makes my entire codebase have to either cast the result to the non-partial value (`T`) or `@ts-ignore` the type in order to get my defined type instead of the deep partial one. 

The `watch` type from the `useFormContext` is unchanged but that always results in infinite updates when I try to use it the same way I use `useWatch` and gives me performance issues so I've moved away from using it even though the typing there is what I need.

**Comment by delyan-peev** [comment]
+1, this is a peculiar case we've been repeatedly fighting in similar ways amongst our codebase

**Comment by zetdotdev** [comment]
@bluebill1049 @Moshyfawn 

Sorry to tag you like this but I really didn't want to open an issue in order to get your attention and I'm pretty sure you guys are the only ones who could answer this well.

**Comment by leapful** [comment]
@Zetxus 

Please provide a codesandbox url to describe more details on your issue. The example from react-hook-form https://codesandbox.io/s/react-hook-form-v7-ts-usewatch-forked-g399v7 shows correct type without casting or ignore anything.

**Reply by delyan-peev** [reply]
OP explicitly mentioned parameterless. I think that's what he meant:

https://codesandbox.io/s/react-hook-form-v7-ts-usewatch-forked-mykjmc?file=/src/index.tsx

Check the different returned types from `useWatch()` and `watch()`

P.S. What's strange to me is that in the case you showed where you `useWatch` a field by name, the return type is correct, but when used without parameters (`useWatch`ing the whole form), the return type is a `DeepPartial`. There is a possible workaround where you place the whole form inside of a root-level field (e.g. `data`) and then use `useWatch({ name: 'data' })` but it seems a bit hacky. Currently, we are having trouble finding a way to leverage the efficiency of `useWatch` when you need to watch the whole form.

**Reply by zetdotdev** [reply]
That's exactly what I meant, my bad for neglecting to provide an example. The provided example displays what I'm experiencing as well.

**Reply by zetdotdev** [reply]
@leapful any insights on this?

**Reply by leapful** [reply]
@Zetxus @delyan-peev 

I think it was caused by different type definition of parameterless `useWatch()` and `watch()`, one is using deep partial, the other does not. The reason is because `useWatch` hook also supports parameter `defaultValue` so it must be deep partial to match the context value.

Could you please open an issue for this improvement ? Feel free to open a related pull request for this case. Thank you :D

**Reply by zetdotdev** [reply]
Sorry we've been crunching for the past few weeks so I just managed to get around to do this. I opened an issue linking to this discussion. If I manage to clear some time I'll try to familiarize myself with the code and open a PR, but I'm not optimistic about my timetable right now.

**Comment by glenjamin** [comment]
I just ran into this too - there was an open issue describing the problem in https://github.com/react-hook-form/react-hook-form/issues/9919 - but it was closed as resolved despite not being resolved - was that intentional?

**Reply by bigsaigon333** [reply]
I'm curious, too 👀
